README.md
# Cache

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/cache/workflows/CI/badge.svg)](https://github.com/originphp/cache/actions)
[![coverage](https://coveralls.io/repos/github/originphp/cache/badge.svg?branch=master)](https://coveralls.io/github/originphp/cache?branch=master)

The OriginPHP Cache library `Apcu`, `Memcached` and `Redis` out of the box, you can also use file caching for large objects or time consuming generating content.

## Installation

To install this package

```linux
$ composer require originphp/cache
```

In your bootstrap file you will the configuration for cache. The caching library can work with multiple configurations and engines at the same time. 



Once the configuration is out the way, using the cache is pretty straightforward. Configure a default Cache store

```php
Cache::config('default', [
    'engine' => 'File',
    'duration' => '+60 minutes', // string or number of seconds e.g. 3600,
    'prefix' => 'cache_'
    'serialize' => true // set to false if you going to cache strings such as output
    'path' => dirname(__DIR__) . '/cache'
     ]);
```

## Caching

### Write

To add an item to the cache.

```php
Use Origin\Cache\Cache;

$success = Cache::write('key',$value);
```

### Read

> From version 3.x when reading from the cache, if there is no hit then it will return `null` instead of `false` as in previous versions.

To read an item from the cache, if it does not find an item it will return `null`

```php
Use Origin\Cache\Cache;

$value = Cache::read('key');
```

### Exists

To check whether a key exists in the cache

```php
Use Origin\Cache\Cache;

if(Cache::exists('key')){
    $bool = Cache::read('key');
}
```

### Delete

Items are automatically deleted based upon the duration setting in the configuration, however if you want
to delete an item manually then use the delete method.

```php
Use Origin\Cache\Cache;

Cache::delete('key');
```


### Clearing the Cache

```php
Cache::clear();
```

### Enabling and disabling the cache

Sometimes you will need to disable the cache, when you disable we switch the engine to the `NullEngine` and your program can work as normal.

```php
Cache::disable();
Cache::enable();
```

### Working with Multiple Configurations

Whether you are using multiple caching engines, or you multiple configurations for a single cache engine (e.g. short duration and long duration caches), the Cache utility is flexible.

You can get the configured Cache store:

```php
$cache = Cache::store('long-duration');
$value = $cache->read('My.key');
```

Or you can pass an options array telling the Cache object which configuration to use


```php
$value = Cache::read('My.key',[
     'config'=>'long-duration'
     ]);
```

## Engines

In all these examples,we are only configuring the default configuration, you can set different configuration names instead of default. When you use the caching functions the default configuration is used by default unless you say otherwise.

### File Engine

```php
Cache::config('default', [
    'engine' => 'File',
    'duration' => '+60 minutes', // string or number of seconds e.g. 3600,
    'prefix' => 'cache_'
    'serialize' => true // set to false if you going to cache strings such as output,
    'mode' => 0664
     ]);
```

### Apcu Engine

```php
Cache::config('default', [
    'engine' => 'Apcu',
    'duration' => '+60 minutes', // string or number of seconds e.g. 3600,
    'prefix' => 'cache_'
     ]);
```

### Memcached Engine

This is a simple configuration for using Memcached.

```php
Cache::config('default', [
        'engine' => 'Memcached',
        'host' => '127.0.0.1',
        'port' => '11211',
        'duration' => '+60 minutes', // string or number of seconds e.g. 3600,
        'prefix' => 'cache_'
     ]);
```

If your Memcached server is configured with username and password then

```php
Cache::config('default', [
        'engine' => 'Memcached',
        'host' => '127.0.0.1',
        'port' => '11211',
        'username' => 'james',
        'password' => 'secret',
        'duration' => '+60 minutes', // 3600,
        'prefix' => 'cache_'
     ]);
```

If you are going to use socket then instead of setting host and port, then set the `path` key with the location
of the socket.

```php
Cache::config('default', [
     'engine' => 'Memcached',
     'path' => '/var/sockets/memcached'
     ]);
```

You can also make connections persistent by setting the `persistent` key to true, or a string which will be the persistent id.

Memcached supports server pools, if you are going to use them then set an array using the `servers` key instead of host and port. The array should be compatible with [memcached addservers](http://php.net/manual/en/memcached.addservers.php).

### Redis Engine

This is a simple configuration for using Redis.

```php
Cache::config('default', [
        'engine' => 'Redis',
        'host' => '127.0.0.1',
        'port' => 6379,
        'duration' => '+60 minutes', // string or number of seconds e.g. 3600,
        'timeout' => 0,
        'prefix' => 'cache_'
     ]);
```

If your Redis server is configured with a password then

```php
Cache::config('default', [
        'engine' => 'Redis',
        'host' => '127.0.0.1',
        'port' =>  6379,
        'password' => 'secret',
        'duration' => 3600, // duration can also be string e.g. +60 minutes
        'prefix' => 'cache_'
     ]);
```

If you are going to use socket then instead of setting host and port, then set the `path` key with the location of the socket.

```php
Cache::config('default', [
        'engine' => 'Redis',
        'path' => '/var/sockets/redis',
     ]);
```

You can also make connections persistent by setting the `persistent` key to true, or a string which will be the persistent id.

```php
Cache::config('default', [
        'engine' => 'Redis',
        'host' => '127.0.0.1',
        'port' => 6379,
        'persistent' => 'my-app',
        'duration' => 3600, // duration can also be string e.g. +60 minutes
        'timeout' => 0,
        'prefix' => 'cache_'
     ]);
```

### Custom Engine

If you want to work with a different backend, it easy to create your own. When configuring cache, instead of passing the `engine` key, use `className` and include the full namespace. 

```php
Cache::config('default', [
    'className' => 'App\Cache\CustomEngine',
    'duration' => 3600,
    'prefix' => 'cache_'
     ]);
```

```php
namespace App\Cache;
use Origin\Cache\Engine\BaseEngine;
class CustomEngine extends BaseEngine
{

}
```

## Installing Cache Engines

The command line instructions have been tested with Ubuntu 18.x.

### Apcu

Apcu is already install in the docker container, however if you need to install this manually.

```php
sudo apt-get update
sudo apt-get install php-apcu
sudo echo 'apc.enable_cli=1' >>  /etc/php/7.2/cli/php.ini
```

### Memcached

**To install Memcached in the Docker container**

First add the following to the `docker-compose.yml` file, this will load the memcached image.

```
  memcached:
      image: memcached
```

In the `Dockerfile` add `php-memcached` to the lines where it installs the php extensions
```
    php-memcached \
```
Then run the build command in docker-compose.

```linux
docker-compose build
```

Then set the host to `memcached` in your cache config.

**To install Memcached on a Ubuntu/Debain based server**

```php
sudo apt-get update
sudo apt-get install memcached
sudo apt-get install php-memcached
```

### Redis


**To install Redis in the Docker container**

First add the following to the `docker-compose.yml` file, this will load the Redis image.

```
  redis:
      image: redis
```

In the `Dockerfile` add the following lines to install and enable the Redis PHP extension.
```
RUN pecl install redis
RUN echo 'extension=redis.so' >> /etc/php/7.2/apache2/php.ini
RUN echo 'extension=redis.so' >> /etc/php/7.2/cli/php.ini
```
Then run the build command in docker-compose.

```linux
docker-compose build
```

Then set the host to `redis` in your cache config.

**To install Redis on a Ubuntu/Debain based server**

```php
pecl install redis
sudo echo 'extension=redis.so' >> /etc/php/7.2/apache2/php.ini
sudo echo 'extension=redis.so' >> /etc/php/7.2/cli/php.ini
```

## Testing

To test this first build the docker container

```linux
docker-compose build
```

To start the container, and access bash

```linux
$ docker-compose up
$ docker-compose run app bash
$ vendor/bin/phpunit
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-01-04

### Changed

- Changed default return from when reading from the cache to NULL from false (BC)
- Changed minimum PHP version to 7.3
- Changed minimum PHPUnit to 9.2

## [1.2.5] - 2020-09-07

### Fixed

- Fixed FileEngine issue in retrieving values when serialization is disabled
- Fixed FileEngine issue when clearing cache with multiple cache configurations using subdirectories

## [1.2.4] - 2020-07-23

### Fixed

- Fixed error handling for incorrect password due to recent changes in extension

## [1.2.3] - 2020-07-16

### Added

- Added FileEngine `mode` for newly created cache files

### Fixed

- Fixed issue with changing duration using strings from individual engines which did not convert to integer

## [1.2.2] - 2020-06-01

### Changed

- Adjusted travis php 7.4 configuration

## [1.2.0] - 2019-10-31

### Changed

- Enabled FileEngine increment / decrement
- FileEngine will create path if it does not exist

## [1.1.0] - 2019-10-30

### Changed

- Changed Redis engine to serialize data and work in a unified way with other engines.

## [1.0.2] - 2019-10-29

### Fixed

- Fixed return types for increment/decrement

## [1.0.1] - 2019-10-14

### Fixed

- Bumped configurable version to 1.x

## [1.0.0] - 2019-10-14

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).

src/Engine/MemcachedEngine.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright    Copyright (c) Jamiel Sharief
 * @link         https://www.originphp.com
 * @license      https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types=1);
namespace Origin\Cache\Engine;

/**
 * Docker Installation
 *  Add to docker-compose.yml - server host will be called 'memached'
 *  memcached:
 *   image: memcached
 * In the docker file add the following line to where it installs exentions etc
 *   php-memcached \
 *
 * For non docker containers:
 * Ubunut/Debian based
 * Instructions to install memcached server and php extension
 * apt-get update
 * apt-get install memcached
 * apt-get install php-memcached
 *
 * Redhat/Fedora based
 * yum update
 * yum install memcached
 * yum install php-memcached
 *
 * If you are working in Docker, then you will need to adjust the docker file and build it again using docker-compose build
 */
use Memcached;
use Origin\Cache\Exception\Exception;

class MemcachedEngine extends BaseEngine
{

    /**
     * Memcached Object
     *
     * @var Memcached
     */
    protected $Memcached = null;

    protected $defaultConfig = [
        'host' => '127.0.0.1',
        'port' => 11211,
        'servers' => [], // if this is defined then a pool is used instead [memached]comptabile with http://php.net/manual/en/memcached.addservers.php
        'username' => null,
        'password' => null,
        'persistent' => false, // set true || string id e.g my-app-xx-, my-app-yyy etc
        'path' => null, // Path to memcached unix socket,
        'duration' => 3600, // memcache has limits if more than 30 days
        'prefix' => 'origin_',
    ];

    /**
     * Constructor
     *
     * @param array $config
     */
    public function initialize(array $config): void
    {
        $msg = 'Memcached extension not loaded.';
        if (extension_loaded('memcached')) {
            $id = null;
            if ($this->config['persistent']) {
                $id = $this->persistentId();
            }
            $this->Memcached = new Memcached($id);
            $msg = 'Error connecting to Memcached server(s).';
            if ($this->connect()) {
                // Login
                if ($this->config['username'] && $this->config['password']) {
                    $this->Memcached->setOption(Memcached::OPT_BINARY_PROTOCOL, true);
                    $this->Memcached->setSaslAuthData($this->config['username'], $this->config['password']);
                }
                if ($this->Memcached->set('---origin-php---', true, 1)) {
                    return;
                }
            }
        }
        throw new Exception($msg);
    }

    protected function connect()
    {
        if (! empty($this->config['servers'])) {
            return $this->Memcached->addServers($this->config['servers']);
        }
        extract($this->config);
        if ($this->config['path']) {
            $host = $this->config['path'];
            $port = 0;
        }

        return $this->Memcached->addServer($host, $port);
    }

    /**
     * Sets a value in the cache
     *
     * @param string $key  max 250 chars
     * @param mixed $value
     * @return bool
     */
    public function write(string $key, $value): bool
    {
        return $this->Memcached->set($this->key($key), $value, $this->duration());
    }
    
    /**
     * Reads a value from the cache, and returns null if there is no hit.
     *
     * @param string $key
     * @return mixed
     */
    public function read(string $key)
    {
        return $this->Memcached->get($this->key($key)) ?: null;
    }

    /**
     * Checks if a key exists in the cache
     *
     * @param string $key
     * @return boolean
     */
    public function exists(string $key): bool
    {
        $this->Memcached->get($this->key($key));

        return ($this->Memcached->getResultCode() === Memcached::RES_SUCCESS);
    }
    /**
     * Deletes a kehy from the cache
     *
     * @param string $key
     * @return boolean
     */
    public function delete(string $key): bool
    {
        return $this->Memcached->delete($this->key($key));
    }

    /**
     * Clears the Cache
     *
     * @return bool
     */
    public function clear(): bool
    {
        return $this->Memcached->flush();
    }

    /**
     * Increases a value
     *
     *  Cache::write('my_value',100);
     *  $value = Cache::increment('my_value');
     *
     * @param string $key
     * @param integer $offset
     * @return integer|bool
     */
    public function increment(string $key, int $offset = 1)
    {
        return $this->Memcached->increment($this->key($key), $offset);
    }

    /**
     * Decreases a value
     *
     * @param string $key
     * @param integer $offset
     * @return integer|bool
     */
    public function decrement(string $key, int $offset = 1)
    {
        return $this->Memcached->decrement($this->key($key), $offset);
    }
}
